ARROW-15911: [C++] Find correct GTest on macOS with Xcode

save-buffer opened the pull request:
Was hitting some issue on Xcode with it not being able to find the correct version of gtest (it's suffixed with "d" if it's debug and release otherwise). I think the issue is that Xcode figures out for itself whether to build Debug or Release _after_ CMake runs, which means that Xcode is stuck only being able to build in one of the two.

save-buffer commented:
@kou I was told you could help review this, would you mind taking a look? Thanks!

pitrou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1844,10 +1844,14 @@ macro(build_gtest)
+    endif()
   else()
-    set(CMAKE_GTEST_DEBUG_EXTENSION "")
+    set(CMAKE_GTEST_DEBUG_EXTENSION "d")
```
I don't understand, why is it always "d" for XCode?

save-buffer replied:
As far as I can tell, it's because Xcode selects whether to build in debug or release mode after configuration, so if it's not set then it will succeed building in Release but not in Debug because it'll try to find `libgtest.dylib` instead of `libgtestd.dylib`. Without this change, if I specified `-DCMAKE_BUILD_TYPE=Debug`, Xcode would be able to build in debug mode but not release mode.

pitrou replied:
Do you mean it always builds GTest in debug mode?

save-buffer replied:
I think it'll build GTest the correct way, but it won't name it properly.

westonpace replied:
Hmm...I get this error in Windows if I do `cmake --build .` instead of `cmake --config Debug --build .`

Coming from Linux this is confusing since I never have to specify `--config` but perhaps the issue here is related?

save-buffer replied:
Could be - what do you use to build on Windows? Visual Studio?

westonpace replied:
Both visual studio and command line (with visual studio generator)

save-buffer replied:
Ah yeah then that might be the same problem. Visual Studio probably also sets the build type independently of configuration.

save-buffer replied:
Any chance you can test out this fix on Windows?

kou replied:
How about using generator expression https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html ? We can use generator expression in `ExternalProject_Add`'s arguments: https://cmake.org/cmake/help/latest/module/ExternalProject.html

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index ff48ac7ee5..44cab1d6b1 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -1849,11 +1849,7 @@ macro(build_gtest)
   set(GTEST_VENDORED TRUE)
   set(GTEST_CMAKE_CXX_FLAGS ${EP_CXX_FLAGS})
 
-  if(CMAKE_BUILD_TYPE MATCHES DEBUG)
-    set(CMAKE_GTEST_DEBUG_EXTENSION "d")
-  else()
-    set(CMAKE_GTEST_DEBUG_EXTENSION "")
-  endif()
+  set(CMAKE_GTEST_DEBUG_EXTENSION "$<$<CONFIG:Debug>,d>")
 
   if(APPLE)
     set(GTEST_CMAKE_CXX_FLAGS ${GTEST_CMAKE_CXX_FLAGS} -DGTEST_USE_OWN_TR1_TUPLE=1
@@ -1894,9 +1890,9 @@ macro(build_gtest)
   set(GTEST_CMAKE_ARGS
       ${EP_COMMON_TOOLCHAIN}
       -DBUILD_SHARED_LIBS=ON
-      -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+      -DCMAKE_BUILD_TYPE=$<CONFIG>
       -DCMAKE_CXX_FLAGS=${GTEST_CMAKE_CXX_FLAGS}
-      -DCMAKE_CXX_FLAGS_${UPPERCASE_BUILD_TYPE}=${GTEST_CMAKE_CXX_FLAGS}
+      -DCMAKE_CXX_FLAGS_$<UPPER_CASE:$<CONFIG>>=${GTEST_CMAKE_CXX_FLAGS}
       -DCMAKE_INSTALL_LIBDIR=lib
       -DCMAKE_INSTALL_NAME_DIR=${GTEST_INSTALL_NAME_DIR}
       -DCMAKE_INSTALL_PREFIX=${GTEST_PREFIX}
```

pitrou commented on cpp/CMakeLists.txt:
```diff
@@ -47,6 +47,10 @@ if(POLICY CMP0074)
 endif()
 
+if(POLICY CMP0114)
+  cmake_policy(SET CMP0114 NEW)
```
Would you mind adding a comment explaining why this is necessary? So that future readers are not confused by this.

save-buffer replied:
Added a comment, I don't know the exact details, just that CMake complains vocally with a lot of yellow text if it's not set (or set to OLD).

save-buffer commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1844,10 +1844,14 @@ macro(build_gtest)
 
-  if(CMAKE_BUILD_TYPE MATCHES DEBUG)
-    set(CMAKE_GTEST_DEBUG_EXTENSION "d")
+  if(NOT CMAKE_GENERATOR STREQUAL Xcode)
```
`GENERATOR_IS_MULTI_CONFIG` seems to work on Xcode for me, I've switched to it. @westonpace could you give it a go on Visual Studio?

save-buffer replied:
As for generator expression, I got a lot of repeating of this:
```
CMake Error at /opt/homebrew/Cellar/cmake/3.22.3/share/cmake/Modules/ExternalProject.cmake:2406:EVAL:2 (add_custom_command):
  Error evaluating generator expression:

    $<$<CONFIG:Debug>,d>

  Expression did not evaluate to a known generator expression
```

Not too sure what to do about it, so I've not switched to generator expression yet.

kou replied:
Ah, sorry. I had a typo: `,` -> `:`: `$<$<CONFIG:Debug>:d>`

westonpace commented:
I'm not sure this change, as it stands, does anything.  But I am no cmake expert.  `GENERATOR_IS_MULTI_CONFIG` is a global property and not a variable.  So, there is no variable named `GENERATOR_IS_MULTI_CONFIG` and so it defaults to `""`.  So it will never affect the if statement because it will always be false.

On my Windows machine it doesn't matter if I'm using the VS generator or the Ninja generator, it still sets `CMAKE_GTEST_DEBUG_EXTENSION` based on `CMAKE_BUILD_TYPE`.

On the other hand, if I do (note this is what we do later on in the line 1941 link @kou provided)...

```
get_property(_GENERATOR_IS_MULTI_CONFIG GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
if(_GENERATOR_IS_MULTI_CONFIG)
```

...then it does recognize the property.  However, if I then try and build in release mode, it fails because it can't find the file.  So all this does is switch multi-config generators from "always fails when building debug" to "always fails when building release".  You claim `Multiconfiguration generators always want the flag set` but I'm not sure that statement is true.  A multi-configuration generator wants `gtestd.dll` if run with `--config Debug` and it wants `gtest.dll` if run with `--config Release`.  There is no way to know, at configuration time, what will be needed at runtime.

So if we are going to build gtest at configuration time (which we need to do if we want to use gtest for test discovery) then we either need to build both versions if it is multi-config or we need to always build one version but only use that version for test discovery and not for a runtime dependency.

westonpace commented:
I'm not sure this change, as it stands, does anything.  But I am no cmake expert.  `GENERATOR_IS_MULTI_CONFIG` is a global property and not a variable.  So, there is no variable named `GENERATOR_IS_MULTI_CONFIG` and so it defaults to `""`.  So it will never affect the if statement because it will always be false.

On my Windows machine it doesn't matter if I'm using the VS generator or the Ninja generator, it still sets `CMAKE_GTEST_DEBUG_EXTENSION` based on `CMAKE_BUILD_TYPE`.

On the other hand, if I do (note this is what we do later on in the line 1941 link @kou provided)...

```
get_property(_GENERATOR_IS_MULTI_CONFIG GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
if(_GENERATOR_IS_MULTI_CONFIG)
```

...then it does recognize the property.  However, if I then try and build in release mode, it fails because it can't find the file.  So all this does is switch multi-config generators from "always fails when building debug" to "always fails when building release".  You claim `Multiconfiguration generators always want the flag set` but I'm not sure that statement is true.  A multi-configuration generator wants `gtestd.dll` if run with `--config Debug` and it wants `gtest.dll` if run with `--config Release`.  There is no way to know, at configuration time, what will be needed at runtime.

So if we are going to build gtest at configuration time (which we need to do if we want to use gtest for test discovery) then we either need to build both versions if it is multi-config or we need to always build one version but only use that version for test discovery and not for a runtime dependency.

kou commented:
(I'm trying to reproduce this on Linux with `-G "Ninja Multi-Config"`. Please wait for a week... I don't have enough time to work on this for now...)

kou commented:
It seems that recent GoogleTest removes `d` postfix: https://github.com/google/googletest/pull/3732
The change isn't released yet. How about using the current main for bundled GoogleTest?
We can change to use a released version again once GoogleTest 1.12.0 is released.

save-buffer commented:
Oh thank you for looking into this! I think that's a great solution too. If we switch to the GTest head, we can just get rid of that whole block dealing with the d suffix right?
Also looking at GTest's readme, I see that it says 
> GoogleTest now follows the [Abseil Live at Head philosophy](https://abseil.io/about/philosophy#upgrade-support)[](https://github.com/apache/arrow/pull/12605#documentation-updates). We recommend using the latest commit in the main branch in your projects.

Maybe we could consider just always using the latest commit?

kou commented:
> If we switch to the GTest head, we can just get rid of that whole block dealing with the d suffix right?

I think so. Could you try it?

> Also looking at GTest's readme, I see that it says
> 
> > GoogleTest now follows the [Abseil Live at Head philosophy](https://abseil.io/about/philosophy#upgrade-support). We recommend using the latest commit in the main branch in your projects.
> 
> Maybe we could consider just always using the latest commit?

Ah, I didn't notice it.
Could you use the latest commit?

save-buffer commented:
Switching to latest commit seems to work for me.

kou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1850,12 +1849,6 @@ macro(build_gtest)
-  else()
-    set(CMAKE_GTEST_DEBUG_EXTENSION "")
-  endif()
-
```
Could you also remove `${CMAKE_GTEST_DEBUG_EXTENSION}` from `_GTEST_LIBRARY_SUFFIX` and `_GTEST_RUNTIME_SUFFIX`?

save-buffer replied:
Done

kou commented on cpp/thirdparty/versions.txt:
```diff
@@ -121,7 +121,7 @@ DEPENDENCIES=(
   "ARROW_GOOGLE_CLOUD_CPP_URL google-cloud-cpp-${ARROW_GOOGLE_CLOUD_CPP_BUILD_VERSION}.tar.gz https://github.com/googleapis/google-cloud-cpp/archive/refs/tags/${ARROW_GOOGLE_CLOUD_CPP_BUILD_VERSION}.tar.gz"
   "ARROW_GRPC_URL grpc-${ARROW_GRPC_BUILD_VERSION}.tar.gz https://github.com/grpc/grpc/archive/${ARROW_GRPC_BUILD_VERSION}.tar.gz"
-  "ARROW_GTEST_URL gtest-${ARROW_GTEST_BUILD_VERSION}.tar.gz https://github.com/google/googletest/archive/release-${ARROW_GTEST_BUILD_VERSION}.tar.gz"
+  "ARROW_GTEST_URL {ARROW_GTEST_BUILD_VERSION}.zip https://github.com/google/googletest/archive/${ARROW_GTEST_BUILD_VERSION}.zip"
```
Could you use `.tar.gz` like others?

save-buffer replied:
Switched it.

save-buffer commented:
There was also a failure with jemalloc in debug builds now after I rebased. It seems that we were appending `--enable-debug` after appending the `EP_LOG_OPTIONS`, which made `log_output_on_failure` be set to `1;--enable-debug`, which caused a failure. I fixed that also

kou commented on cpp/thirdparty/versions.txt:
```diff
@@ -121,7 +121,7 @@ DEPENDENCIES=(
   "ARROW_GOOGLE_CLOUD_CPP_URL google-cloud-cpp-${ARROW_GOOGLE_CLOUD_CPP_BUILD_VERSION}.tar.gz https://github.com/googleapis/google-cloud-cpp/archive/refs/tags/${ARROW_GOOGLE_CLOUD_CPP_BUILD_VERSION}.tar.gz"
   "ARROW_GRPC_URL grpc-${ARROW_GRPC_BUILD_VERSION}.tar.gz https://github.com/grpc/grpc/archive/${ARROW_GRPC_BUILD_VERSION}.tar.gz"
-  "ARROW_GTEST_URL gtest-${ARROW_GTEST_BUILD_VERSION}.tar.gz https://github.com/google/googletest/archive/release-${ARROW_GTEST_BUILD_VERSION}.tar.gz"
+  "ARROW_GTEST_URL {ARROW_GTEST_BUILD_VERSION}.tar.gz https://github.com/google/googletest/archive/${ARROW_GTEST_BUILD_VERSION}.tar.gz"
```
Could you add `googletest-` prefix for easy to know package name:

```console
$ curl -L -v https://github.com/google/googletest/archive/af29db7ec28d6df1c7f0f745186884091e602e07.tar.gz 2>&1 | grep filename= 
< content-disposition: attachment; filename=googletest-af29db7ec28d6df1c7f0f745186884091e602e07.tar.gz
```

save-buffer replied:
Just adding `googletest-` is sufficient right? I've added that back in

kou commented:
@westonpace do you want to try this on Windows before we merge this?

westonpace reviewed (approved):
Unfortunately, I had already pinned gtest to 1.10.0 in my Windows builds to avoid unresolved external errors.  However, that is a separate issue (that I guess I'll go debug now :) ).  I did confirm that this version finds and copies the dll correctly for both debug and release builds on Windows.

westonpace commented:
The appveyor failure looks like a problem.  It's a slightly different link error than the one I'm getting but I think in spirit it is the same.  The arrow testing library gets built as a shared object and references some of gtest.  Those parts of gtest do not seem to be properly exporting their symbols.

amol- commented:
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍
